presto-main/src/main/java/com/facebook/presto/memory/RemoteNodeMemory.java

```diff
@@ -0,0 +1,93 @@
+import io.airlift.http.client.FullJsonResponseHandler.JsonResponse;
+import io.airlift.http.client.HttpClient;
+import io.airlift.http.client.HttpClient.HttpResponseFuture;
+import io.airlift.http.client.Request;
+import io.airlift.json.JsonCodec;
+import io.airlift.log.Logger;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.net.URI;
+import java.util.Optional;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static io.airlift.http.client.FullJsonResponseHandler.createFullJsonResponseHandler;
+import static io.airlift.http.client.Request.Builder.prepareGet;
+import static io.airlift.units.Duration.nanosSince;
+import static java.util.Objects.requireNonNull;
+import static java.util.Optional.empty;
+import static java.util.Optional.ofNullable;
+
+@ThreadSafe
+public class RemoteNodeMemory
+{
+    private static final Logger log = Logger.get(RemoteNodeMemory.class);
+
+    private final HttpClient httpClient;
+    private final URI memoryInfoUri;
+    private final JsonCodec<MemoryInfo> memoryInfoCodec;
+    private final AtomicReference<Optional<MemoryInfo>> memoryInfo = new AtomicReference<>(empty());
+    private final AtomicReference<Future<?>> future = new AtomicReference<>();
+    private final AtomicLong lastUpdateNanos = new AtomicLong();
+
+    public RemoteNodeMemory(HttpClient httpClient, JsonCodec<MemoryInfo> memoryInfoCodec, URI memoryInfoUri)
+    {
+        this.httpClient = requireNonNull(httpClient, "httpClient is null");
+        this.memoryInfoUri = requireNonNull(memoryInfoUri, "memoryInfoUri is null");
+        this.memoryInfoCodec = requireNonNull(memoryInfoCodec, "memoryInfoCodec is null");
+    }
+
+    public Optional<MemoryInfo> getInfo()
+    {
+        return memoryInfo.get();
+    }
+
+    public void asyncRefresh()
+    {
+        if (nanosSince(lastUpdateNanos.get()).toMillis() > 1_000 && future.get() == null) {
+            Request request = prepareGet().setUri(memoryInfoUri).build();
+            HttpResponseFuture<JsonResponse<MemoryInfo>> responseFuture = httpClient.executeAsync(request, createFullJsonResponseHandler(memoryInfoCodec));
+            future.set(responseFuture);
+
+            Futures.addCallback(responseFuture, new FutureCallback<JsonResponse<MemoryInfo>>() {
+                @Override
+                public void onSuccess(@Nullable JsonResponse<MemoryInfo> result)
+                {
+                    if (result != null && result.hasValue()) {
+                        memoryInfo.set(ofNullable(result.getValue()));
+                    }
+                    lastUpdateNanos.set(System.nanoTime());
+                    future.set(null);
+                }
+
+                @Override
+                public void onFailure(Throwable t)
+                {
+                    log.warn(t, "Error fetching memory info from %s", memoryInfoUri);
```

Comment:
If there is a problem with a node, will this fill the log?  If so, add an `AtomicBoolean` to only log when you go from good to bad and bad to good.